File: android/core-domain/src/main/kotlin/com/crispy/tv/domain/browse/BrowseTitlesPlanner.kt
package com.crispy.tv.domain.browse

data class BrowseCombo(
    val type: String,
    val genre: String?,
    val sort: String,
)

data class BrowseRequest(
    val type: String,
    val genre: String?,
    val sort: String,
    val page: Int,
)

data class BrowseTypeItem(
    val itemId: String,
    val type: String,
)

data class BrowseTypeResponse(
    val items: List<BrowseTypeItem> = emptyList(),
    val hasMore: Boolean = false,
)

data class BrowsePageResult(
    val itemIds: List<String> = emptyList(),
    val hasMore: Boolean = false,
    val nextPage: Int? = null,
)

const val BROWSE_MAX_PAGES = 20

fun planBrowseRequests(combo: BrowseCombo, page: Int): List<BrowseRequest> {
    if (page < 0 || page >= BROWSE_MAX_PAGES) return emptyList()
    val genre = normalizedBrowseGenre(combo.genre)
    return browseTypeFanOut(combo.type).map { type ->
        BrowseRequest(
            type = type,
            genre = genre,
            sort = combo.sort,
            page = page,
        )
    }
}

fun mergeBrowsePage(
    combo: BrowseCombo,
    page: Int,
    responses: Map<String, BrowseTypeResponse>,
): BrowsePageResult {
    if (page < 0 || page >= BROWSE_MAX_PAGES) {
        return BrowsePageResult()
    }

    val types = browseTypeFanOut(combo.type)
    val itemIds =
        types.flatMap { type ->
            responses[type]?.items.orEmpty().map { it.itemId }
        }
    val hasMore =
        types.any { type -> responses[type]?.hasMore ?: false } && page < BROWSE_MAX_PAGES - 1
    return BrowsePageResult(
        itemIds = itemIds,
        hasMore = hasMore,
        nextPage = if (hasMore) page + 1 else null,
    )
}

private fun browseTypeFanOut(type: String): List<String> {
    return if (type == "all") listOf("movie", "series") else listOf(type)
}

private fun normalizedBrowseGenre(genre: String?): String? {
    return genre?.trim()?.takeIf { it.isNotEmpty() }
}

File: ios/ContractRunner/Sources/ContractRunner/BrowseTitlesPlanner.swift
import Foundation

public struct BrowseCombo: Equatable {
    public let type: String
    public let genre: String?
    public let sort: String

    public init(type: String, genre: String?, sort: String) {
        self.type = type
        self.genre = genre
        self.sort = sort
    }
}

public struct BrowseRequest: Equatable {
    public let type: String
    public let genre: String?
    public let sort: String
    public let page: Int

    public init(type: String, genre: String?, sort: String, page: Int) {
        self.type = type
        self.genre = genre
        self.sort = sort
        self.page = page
    }
}

public struct BrowseTypeItem: Equatable {
    public let itemId: String
    public let type: String

    public init(itemId: String, type: String) {
        self.itemId = itemId
        self.type = type
    }
}

public struct BrowseTypeResponse: Equatable {
    public let items: [BrowseTypeItem]
    public let hasMore: Bool

    public init(items: [BrowseTypeItem] = [], hasMore: Bool = false) {
        self.items = items
        self.hasMore = hasMore
    }
}

public struct BrowsePageResult: Equatable {
    public let itemIds: [String]
    public let hasMore: Bool
    public let nextPage: Int?

    public init(itemIds: [String] = [], hasMore: Bool = false, nextPage: Int? = nil) {
        self.itemIds = itemIds
        self.hasMore = hasMore
        self.nextPage = nextPage
    }
}

public let browseMaxPages = 20

public func planBrowseRequests(combo: BrowseCombo, page: Int) -> [BrowseRequest] {
    guard page >= 0, page < browseMaxPages else {
        return []
    }
    let genre = normalizedBrowseGenre(combo.genre)
    return browseTypeFanOut(combo.type).map { type in
        BrowseRequest(type: type, genre: genre, sort: combo.sort, page: page)
    }
}

public func mergeBrowsePage(
    combo: BrowseCombo,
    page: Int,
    responses: [String: BrowseTypeResponse]
) -> BrowsePageResult {
    guard page >= 0, page < browseMaxPages else {
        return BrowsePageResult()
    }

    let types = browseTypeFanOut(combo.type)
    let itemIds = types.flatMap { type in
        responses[type]?.items.map(\.itemId) ?? []
    }
    let hasMore = types.contains(where: { responses[$0]?.hasMore ?? false }) && page < browseMaxPages - 1
    return BrowsePageResult(
        itemIds: itemIds,
        hasMore: hasMore,
        nextPage: hasMore ? page + 1 : nil
    )
}

private func browseTypeFanOut(_ type: String) -> [String] {
    type == "all" ? ["movie", "series"] : [type]
}

private func normalizedBrowseGenre(_ genre: String?) -> String? {
    guard let trimmed = genre?.trimmingCharacters(in: .whitespacesAndNewlines), !trimmed.isEmpty else {
        return nil
    }
    return trimmed
}